### ISSUE TITLE
Configure Provider Before Cross-Providers Resource Move

### ISSUE THREAD

**Issue by ms-henglu** [hashicorp/terraform#35922]
### Terraform Version

```shell
Terraform v1.9.8
on darwin_arm64
```

### Use Cases

The MoveResourceState protocol (`moved` block) is very useful in cross-provider, but it could only be called before ConfigureProvider, so in the resource move implementation, we could not make API calls.

### Attempted Solutions

NA

### Proposal

Is it possible to make it be called after ConfigureProvider?

**Comment by jbardin** [comment]
Hi @ms-henglu,

Do you have a specific use-case in mind which requires configuring the provider?

State refactoring operations are designed to be executed offline, where there is no access to any provider configuration. Changing how refactoring works entirely would be a large redesign, so there must be a sufficiently compelling use case to warrant the engineering effort. So far providers have been able to complete any necessary transformations to the extent that the resource can be read from its new state and configuration, allowing the provider to complete any actions that need API access during the following plan+apply cycle.

**Comment by ms-henglu** [comment]
Hi @jbardin ,

Thanks for the prompt response and clarification.

Yes, I'm maintaining the azapi provider which is a generic azure provider and I'm trying to use this functionality to achieve resource-move from the azurerm provider. To simplify the implementation, it would be much easier to call the API and convert the API response to the terraform state instead of converting azurerm's resource state into azapi's.

**Comment by jbardin** [comment]
Thanks for the extra context @ms-henglu. The `moved` operation was implemented the way it was for various reasons, one of which is that `import` alone is not always sufficient to obtain all the necessary information to build an instance value, and `MoveResourceState` was implemented so that providers could have access to all the current data stored in each resource instance.

Most of the import process happens during the `ReadResource` call anyway, so there's really no reason you couldn't get the same behavior as you would with the combination of `removed` and `import`, except you would be starting with more information. As long as the new resource is create with enough information in the state to be read during the plan, you can use the API calls to fill in the rest.

**Comment by ms-henglu** [comment]
Thanks @jbardin

>  Most of the import process happens during the ReadResource call anyway, so there's really no reason you couldn't get the same behavior as you would with the combination of removed and import, except you would be starting with more information. As long as the new resource is created with enough information in the state to be read during the plan, you can use the API calls to fill in the rest.

Yes, but the only difference is when running `terraform plan` command, the `moved` block solution will still have an update.

Please check the below example which migrates the data factory resource from azurerm provider to azapi provider, and let's presume the `publicNetworkAccess` feature was not supported from the source, so its value must be retrieved from API.


1. The `moved` block solution:  
 There's an update action in the plan output, which is kind of misleading, because the resource's remote state is matched with the configuration, only the terraform state is not updated after the `MoveState` phase.


```hcl

# resource "azurerm_data_factory" "test" {
#   name                   = "acctesthenglu1031"
#   location               = azurerm_resource_group.test.location
#   resource_group_name    = azurerm_resource_group.test.name
# }

moved {
  from = azurerm_data_factory.test
  to = azapi_resource.factory
}

resource "azapi_resource" "factory" {
  type      = "Microsoft.DataFactory/factories@2018-06-01"
  parent_id = azurerm_resource_group.test.id
  name      = "acctesthenglu1031"
  location  = "eastus"
  body = {
    properties = {
      globalParameters    = {}
      publicNetworkAccess = "Enabled"  
    }
  }
}
```

```shell
  # azapi_resource.factory will be updated in-place
  # (moved from azurerm_data_factory.test)
  ~ resource "azapi_resource" "factory" {
      + body                      = {
          + properties = {
              + globalParameters    = {}
              + publicNetworkAccess = "Enabled"
            }
        }
        id                        = "/subscriptions/.../resourceGroups/acctesthenglu1031/providers/Microsoft.DataFactory/factories/acctesthenglu1031"
        name                      = "acctesthenglu1031"
        # (6 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

2. The `removed` + `import` block solution:  
The plan output only has one resource to import and one resource is no longer managed by Terraform, which is much cleaner.

```hcl

# resource "azurerm_data_factory" "test" {
#   name                   = "acctesthenglu1031"
#   location               = azurerm_resource_group.test.location
#   resource_group_name    = azurerm_resource_group.test.name
# }

removed {
  from = azurerm_data_factory.test
  lifecycle {
    destroy = false
  }
}

import {
  id = "/subscriptions/.../resourceGroups/acctesthenglu1031/providers/Microsoft.DataFactory/factories/acctesthenglu1031"
  to = azapi_resource.factory
}

resource "azapi_resource" "factory" {
  type      = "Microsoft.DataFactory/factories@2018-06-01"
  parent_id = azurerm_resource_group.test.id
  name      = "acctesthenglu1031"
  location  = "eastus"
  body = {
    properties = {
      globalParameters    = {}
      publicNetworkAccess = "Enabled"
    }
  }
}


```



```shell
Terraform will perform the following actions:

  # azapi_resource.factory will be imported
    resource "azapi_resource" "factory" {
        body                      = {
            properties = {
                globalParameters    = {}
                publicNetworkAccess = "Enabled"
            }
        }
        id                        = "/subscriptions/.../resourceGroups/acctesthenglu1031/providers/Microsoft.DataFactory/factories/acctesthenglu1031"
        ignore_casing             = false
        ignore_missing_property   = true
        location                  = "eastus"
        name                      = "acctesthenglu1031"
        parent_id                 = "/subscriptions/.../resourceGroups/acctesthenglu1031"
        schema_validation_enabled = true
        type                      = "Microsoft.DataFactory/factories@2018-06-01"
    }

 # azurerm_data_factory.test will no longer be managed by Terraform, but will not be destroyed
 # (destroy = false is set in the configuration)
 . resource "azurerm_data_factory" "test" {
        id                               = "/subscriptions/.../resourceGroups/acctesthenglu1031/providers/Microsoft.DataFactory/factories/acctesthenglu1031"
        name                             = "acctesthenglu1031"
        # (7 unchanged attributes hidden)
    }

Plan: 1 to import, 0 to add, 0 to change, 0 to destroy.
```

In summary, I would expect the `moved` block could have same plan output as the `removed`+`import` solution, and configure provider before Cross-Providers resource move could be a solution.

**Comment by jbardin** [comment]
Moves must be completed before we can start evaluating the configuration, which is a prerequisite for configuring the provider, so it's unlikely that will be a possible solution.

The change in the plan indicates that the `body` attribute isn't present after the move has taken place. It would useful to know why your `ReadResource` call is filling in the `body` attribute during import, but not after the move.

**Comment by jbardin** [comment]
It's still not clear why the resource can't be fully recreated after moving it while it can after import. The `ImportResourceState` call typically does little to nothing due to it's legacy implementation, and just transfers the ID to a correctly-typed, but mostly empty value to be filled in by `ReadResource`. If something is missing after the offline move operation, in theory it should be filled in by the same process. In fact stripping out any questionable state data may be preferable in some cases, because of the Azure provider's heavy reliance on legacy SDK behavior which will confuse empty and null values among other things.

If there is some situation in which a different behavior is necessary, you also have the private data field to transfer non-state information between calls.

**Comment by ms-henglu** [comment]
Oh, I guess I didn't make myself clear. Actually the `moved` block and the whole process did work. However it will have the terraform plan output below. **I would expect to see a `no changes` output.** Because to end users, the moved configuration did match with remote state, there's nothing changed(But I understand that the state after read resource does have differences comparing with the state after moving).

```
 # azapi_resource.factory will be updated in-place
  # (moved from azurerm_data_factory.test)
  ~ resource "azapi_resource" "factory" {
      + body                      = {
          + properties = {
              + globalParameters    = {}
              + publicNetworkAccess = "Enabled"
            }
        }
        id                        = "/subscriptions/.../resourceGroups/acctesthenglu1031/providers/Microsoft.DataFactory/factories/acctesthenglu1031"
        name                      = "acctesthenglu1031"
        # (6 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

> It's still not clear why the resource can't be fully recreated after moving it while it can after import.

For example, an azure resource supports `foo` and `bar`. But the source terraform resource type only supports `foo`. When moving it to a target resource type which supports `foo` and `bar`, the value of `bar` could not be generated only based on the source terraform state. But later in the `ReadResource`, the value of `bar` could be retrieved from API.

**Comment by jbardin** [comment]
Sorry, I'm not really understanding what you're presenting here. 

If we're talking about your hypothetical example, yes the "bar" attribute can be filled in by `ReadResource`, both after `ImportResourceState` or `MoveResourceState`.

I assume you are trying to translate idea that to the `body` attribute, which was not supported in the old resource, but can be read from the API in the new resource. If your plan is showing that `body` will be added, it is because it exists in the configuration but not in the state. That means your `ReadResource` did _not_ return the `body` attribute in the updated state for whatever reason. 

If `body` exists in the remote API, you need to ensure that it's being returned when that resource is read. I can't tell from Terraform's side of the transaction why that value is not present, but I can assure you that I tested this workflow internally within Terraform, and both moved and imported resources should be updated identically based on the remote state.

**Comment by ms-henglu** [comment]
Hi @jbardin , 

>  That means your ReadResource did not return the body attribute in the updated state for whatever reason.
You're right, my implementation is incorrect. Thank you so much for the patient clarification and apologize for wasting your time.
